### DISCUSSION TITLE
Architecture and Design Patterns Questions

### DISCUSSION THREAD

**Discussion by lucas1011100** [google/model-viewer#4726] [Q&A]
I am planning to analyze this source, and I was wondering if I could ask which architecture and design patterns were used? Also, I'm curious about the reasons for adopting that architecture.

**Comment by elalish** [comment]
Well, mostly it's a Lit-based web component, so we're doing our best to follow their best practices - everything is async. It uses Mixins, but this decision was made early on when it was thought those pieces might actually be used separately and tree-shaken. In fact, we're quite monolithic, so I find that architecture to be a bit over-engineered. I would remove it if it wasn't such a lot of work, so we make do with just not adding any more of them. 

As with many sizable projects, the architecture is less about named patterns and more about the path of least resistance, and occasionally doing a refactor to clean up when the code has become too tangled. `ModelScene` could really use a refactor as it has become the nexus of everything and has grown too large. 

The most important architecture decision from a performance standpoint is that we have a single shared renderer (static) that every MV on the page flows through. There is quite a bit of complexity around handling resizing and lazy loading properly.
